pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+// Package eventhandler watches for Kubernetes Event objects and hands them off to
+// Agent's Logs subsystem (embedded promtail)
+package eventhandler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	cacheFileMode = 0600
+)
+
+// EventHandler watches for Kubernetes Event objects and hands them off to
+// Agent's logs subsystem (embedded promtail).
+type EventHandler struct {
+	LokiClient    *logs.Instance
+	Log           log.Logger
+	CacheFile     *os.File
+	CachePath     string
+	LastEvent     *ShippedEvents
+	InitEvent     *ShippedEvents
+	EventInformer cache.SharedIndexInformer
+	ClusterLabel  string
+	SendTimeout   time.Duration
+	sync.Mutex
+}
+
+// ShippedEvents stores a timestamp and map of event ResourceVersions shipped for that timestamp.
+// Used to avoid double-shipping events upon restart.
+type ShippedEvents struct {
+	// shipped event's timestamp
+	Timestamp time.Time `json:"ts"`
+	// map of event RVs (resource versions) already "shipped" (handed off) for this timestamp.
+	// this is to handle the case of a timestamp having multiple events,
+	// which happens quite frequently.
+	RvMap map[string]struct{} `json:"resourceVersion"`
+}
+
+func newEventHandler(l log.Logger, globals integrations.Globals, c *Config) (integrations.Integration, error) {
+	var (
+		config *rest.Config
+		err    error
+	)
+
+	if c.InCluster {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags("", c.KubeconfigPath)
+	}
```

Comment:
You could get rid of the `InCluster` option on `Config` here since `BuildConfigFromFlags` will fall back to using `InClusterConfig` if the kubeconfig path is empty.

Reply:
Nice catch, I removed the config option but wanted to preserve checking for a kubeconfig in the user's home dir in case they didn't set `kubeconfig_path`. Let me know what you think of the updated approach!

Reply:
looks good to me!